src/datanode/src/sql/copy_table_from.rs

```diff
@@ -0,0 +1,461 @@
+    pub fn new(
+        uri: String,
+        pattern: Option<String>,
+        connection: HashMap<String, String>,
+    ) -> Result<DataSource> {
+        let result = SCHEMA_PATTERN.find(&uri);
+        let parsed_uri = Uri::parse(&uri).context(error::InvalidUriSnafu {
+            uri: uri.to_string(),
+        })?;
+        let host = parsed_uri
+            .host_to_string()
+            .context(error::InvalidUriSnafu {
+                uri: uri.to_string(),
+            })?;
+
+        let path = parsed_uri
+            .path_to_string()
+            .context(error::InvalidUriSnafu {
+                uri: uri.to_string(),
+            })?;
+
+        let (path, filename) = DataSource::find_path_and_filename(&path);
+        let source = Source::new(filename, pattern)?;
+
+        let object_store = match result {
+            Some(result) => match result.as_str().to_ascii_uppercase().as_str() {
+                S3_SCHEMA => {
+                    let mut builder = S3::default();
+
+                    builder.root(&path);
+
+                    if let Some(bucket) = host {
+                        builder.bucket(&bucket);
+                    }
+
+                    if let Some(endpoint) = connection.get(ENDPOINT_URL) {
+                        builder.endpoint(endpoint);
+                    }
+
+                    if let Some(region) = connection.get(REGION) {
+                        builder.region(region);
+                    }
+
+                    if let Some(key_id) = connection.get(ACCESS_KEY_ID) {
+                        builder.access_key_id(key_id);
+                    }
+
+                    if let Some(key) = connection.get(SECRET_ACCESS_KEY) {
+                        builder.secret_access_key(key);
+                    }
+
+                    if let Some(session_token) = connection.get(SESSION_TOKEN) {
+                        builder.security_token(session_token);
+                    }
+
+                    if let Some(enable_str) = connection.get(ENABLE_VIRTUAL_HOST_STYLE) {
+                        let enable = enable_str.as_str().parse::<bool>();
+
+                        match enable {
+                            Ok(true) => {
+                                builder.enable_virtual_host_style();
+                            }
+                            Err(_) => {
+                                return Err(error::Error::InvalidConnection {
+                                    msg: format!(
+                                        "failed to parse the {}",
+                                        ENABLE_VIRTUAL_HOST_STYLE
+                                    ),
+                                })
+                            }
+                            _ => (),
+                        }
+                    }
+
+                    let accessor = builder.build().context(error::BuildBackendSnafu)?;
+
+                    ObjectStore::new(accessor).finish()
+                }
+                &_ => {
```

Comment:
style: You don't need to write this `&` in Rust
```suggestion
                _ => {
```


See https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html